### PR TITLE
F64 support

### LINKED ISSUE #891: float64 support
Support for float64 over all plattforms.
This would also fix up to 20 ONNX tests.

Is there a plan to add this? Otherwise there is no way to fully support ONNX, since some tests are hard relying on that.

### PR DESCRIPTION
**Pull request by dc-dc-dc**
Adds float64 support to the different runtimes, also included an excluded field on dtypes for unsupported platforms

From my testing float64 is not supported on metal and OpenCL on M2 Mac. OpenCL float64 is fine on my windows machine although this is an optional feature and ideally should check the OpenCL extensions for it.

Also brings the onnx stat line to:
190 failed, 650 passed, 1794 skipped, 2 warnings in 14.73s

Tested on:
- CPU
- GPU
- CUDA
- LLVM
- CLANG

Closes #891

### CODE REVIEW

**Comment by dc-dc-dc** [comment]
Will add the following in the kernel prefix
```
#ifdef cl_khr_fp64
#pragma OPENCL EXTENSION cl_khr_fp64 : enable
#elif defined(cl_amd_fp64)
#pragma OPENCL EXTENSION cl_amd_fp64 : enable
#endif
```

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -36,6 +36,7 @@ class Tensor:
 
   def __init__(self, data:Union[int, float, list, tuple, LazyBuffer, np.ndarray], device=Device.DEFAULT, dtype:Optional[DType]=None, requires_grad:Optional[bool]=None):
     assert dtype is None or isinstance(dtype, DType), f"invalid dtype {dtype}"
+    assert dtype is None or device not in dtype.excluded, f"dtype: {dtype} not supported on device: {device}"
```
Comment (geohot) [inline tinygrad/tensor.py:39]: You want this after the device `canonicalize` and you only want to check before the colon.

"gpu:0", "GPU:0" can both be passed into device

File: tinygrad/runtime/ops_gpu.py [hunk 2]
```diff
@@ -87,7 +86,7 @@ def __call__(self, global_size, local_size, *bufs, wait=False) -> Optional[float
 
 class CLCodegen(CStyleCodegen):
   lang = CStyleLanguage(
-    kernel_prefix = "#define int64 long\n__kernel", buffer_prefix = "__global ", smem_prefix = "__local ",
+    kernel_prefix = "#ifdef cl_khr_fp64\n#pragma OPENCL EXTENSION cl_khr_fp64 : enable\n#elif defined(cl_amd_fp64)\n#pragma OPENCL EXTENSION cl_amd_fp64 : enable\n#endif\n#define int64 long\n__kernel", buffer_prefix = "__global ", smem_prefix = "__local ",
```
Comment (geohot) [inline tinygrad/runtime/ops_gpu.py:89]: Want to do something similar to half_prekernel so this isn't in all the kernels?

File: tinygrad/helpers.py [hunk 3]
```diff
@@ -69,14 +72,15 @@ class dtypes:
...
-  def is_float(x: DType) -> bool: return x in (dtypes.float16, dtypes.float32)
+  def is_float(x: DType) -> bool: return x in (dtypes.float16, dtypes.float32, dtypes.float64)
   @staticmethod
   def is_unsigned(x: DType) -> bool: return x in (dtypes.uint8)
   @staticmethod
   def from_np(x) -> DType: return asdict(dtypes())[np.dtype(x).name]
   bool: Final[DType] = DType(0, 1, "bool", bool)
   float16: Final[DType] = DType(0, 2, "half", np.float16)
   float32: Final[DType] = DType(4, 4, "float", np.float32)
+  float64: Final[DType] = DType(5, 8, "double", np.float64, ("METAL", "GPU" if OSX else ""))
```
Comment (geohot) [inline tinygrad/helpers.py:83]: I don't think you want "" here if not OSX, I think you just don't want that element in the tuple

Comment (dc-dc-dc) [reply tinygrad/helpers.py:83]: yeah will change to `("METAL", "GPU") if OSX else ()`

File: tinygrad/runtime/ops_gpu.py (hunk 2 above)
Comment (dc-dc-dc) [reply tinygrad/runtime/ops_gpu.py:89]: misread this, ignore the last commit

File: tinygrad/tensor.py [hunk 4]
```diff
@@ -37,6 +37,7 @@ class Tensor:
   def __init__(self, data:Union[int, float, list, tuple, LazyBuffer, np.ndarray], device=Device.DEFAULT, dtype:Optional[DType]=None, requires_grad:Optional[bool]=None):
     assert dtype is None or isinstance(dtype, DType), f"invalid dtype {dtype}"
     device = Device.canonicalize(device)
+    assert dtype is None or device.split(":", 1)[0] not in dtype.excluded, f"dtype: {dtype} not supported on device: {device}"
```
Comment (geohot) [inline tinygrad/tensor.py:40]: These asserts should probably be at a much lower level, like when you create the rawbuffer. Then you only need one assert, and you don't need the excluded on dtypes either.

**Comment by geohot** [comment]
Thanks! Now those onnx tests should pass

**Comment by rayanht** [comment]
Should probably add a pytest decorator to skip the float64 tests on Mac since it's not supported

**Comment by jdlib** [comment]
As @dc-dc-dc mentioned initially:

> OpenCL float64 is fine on my windows machine although this is an optional feature and ideally should check the OpenCL extensions for it.

I ran on Windows with a Intel Iris XE card and it fails the TestFloat64Dtype tests
```
error: use of type 'double' requires cl_khr_fp64 extension to be enabled
```
The generated cl file starts as expected, e.g. with
```
#ifdef cl_khr_fp64
#pragma OPENCL EXTENSION cl_khr_fp64 : enable
#elif defined(cl_amd_fp64)
#pragma OPENCL EXTENSION cl_amd_fp64 : enable
#endif
__kernel void E_4(__global double* data0, const __global double* data1, const __global double* data2) {
```
but the extensions of that card only contain `cl_khr_fp16` but not `cl_khr_fp64`.
Also since the error message is somewhat misleading, would it be of advantage to add a #else branch to the generated code?
```
...
#else
#error "fp64 not supported by device"
#endif
````
